### PR TITLE
enhanced type safety 👊

### PR DESCRIPTION
**Pull request by kotarella1110**
CSB: https://codesandbox.io/s/nervous-wind-s7u3w?file=/src/Form.tsx

Usage:

`global.d.ts`:

```ts
declare module 'little-state-machine' {
  interface GlobalState {
    yourDetail: {
      name: string;
    };
  };
}
```

`App.tsx`:

```tsx
import React from 'react';
import {
  createStore,
  useStateMachine,
  StateMachineProvider,
  GlobalState,
} from 'little-state-machine';

const initialState = {
  yourDetail: { name: '' },
};
createStore(initialState);
createStore({ error: '' }); // type error

const updateName = (state: GlobalState, payload: { name: string }) => ({
  ...state,
  yourDetail: {
    ...state.yourDetail,
    ...payload,
  },
});

const YourComponent = () => {
  const { actions, state } = useStateMachine({
    updateName
  }));

  actions.updateName({ error: 'bill' }); // type error
  actions.updateName({ name: 0 }); // type error
  console.log(state.error); // type error
  console.log(state.yourDetail.error); // type error

  return (
    <div onClick={() => actions.updateName({ name: 'bill' })}>
      {state.yourDetail.name}
    </div>
  );
};

const App = () => (
  <StateMachineProvider>
    <YourComponent />
  </StateMachineProvider>
);
```

### CODE REVIEW

**Comment by kotarella1110** [comment]
@bluebill1049 Which global state type naming do you prefer?

- GlobalState
- RootState
- AppState

**Comment by bluebill1049** [comment]
> @bluebill1049 Which global state type naming do you prefer?
> 
> * GlobalState
> * RootState
> * AppState

`GlobalState` sounds great

**Comment by devrnt** [comment]
You could use `typeof` as wel in your module declaration:
```ts
const yourDetailState = {
  name: ""
};

declare module "./src" {
  interface GlobalState {
    yourDetail: typeof yourDetailState;
  }
```

This allows to declare both an initial state and your types.

**Comment by kotarella1110** [comment]
@devrnt It's been such a long time! awesome 👏

File: src/StateMachineContext.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  const value = React.useContext<StateMachineContextValue>(StateMachineContext);
+
+  if (process.env.NODE_ENV !== 'production') {
+    if (!value) {
+      console.error(`StateMachine context is undefined, please verify you are calling useStateMachine() as child of a <StateMachineProvider> component.`)
+    }
+  }
+
+  return value;
+}
```
Comment (bluebill1049) [inline src/StateMachineContext.tsx:29]: thanks for making it into an own file.

File: src/index.ts [hunk 2]
```diff
@@ -1,7 +1,10 @@
 import {
-  createStore,
   StateMachineProvider,
+} from './StateMachineContext';
+import {
+  createStore,
   useStateMachine,
 } from './stateMachine';
+import { GlobalState } from './types';
```
Comment (bluebill1049) [inline src/index.ts:8]: 👁️ 👃 👁️

File: src/logic/getBrowserStoreData.ts [hunk 3]
```diff
@@ -2,6 +2,6 @@ export default (storageType: Storage, storeName: string) => {
   try {
     return JSON.parse(storageType.getItem(storeName) as string);
   } catch {
-    return '';
+    return null;
```
Comment (bluebill1049) [inline src/logic/getBrowserStoreData.ts:5]: 👍

File: src/stateMachine.tsx [hunk 4]
```diff
@@ -23,97 +22,66 @@ export function createStore<T>(
...
 
-function actionTemplate<T>(
-  updateStore: React.Dispatch<T>,
-  callback: StoreUpdateFunction<T>,
+function actionTemplate<TCallback extends AnyCallback>(
+  setState: React.Dispatch<React.SetStateAction<GlobalState>>,
+  callback: TCallback,
 ) {
-  return <K extends DeepPartial<T>>(payload: K) => {
+  return (payload: Parameters<TCallback>[1]) => {
```
Comment (bluebill1049) [inline src/stateMachine.tsx:40]: 👍  I got stuck on this without `Parameters`.

**Comment by Flavien-Pensato** [comment]
I'm not a fan of typing specially at that place. It isn't the store responsibility to check Type.

If I want sometime a string or a array can I do ?

**Comment by bluebill1049** [comment]
@Flavien-Pensato which place? can you give an example.

**Comment by Flavien-Pensato** [comment]
No concrete example to show you sorry :/ The simple fact to know I cannot do anything I want feel sad.

Does the rule apply for field object ?

```
{ user: :{ name: "" }}

{ user: :{ name: 12456 }} // error ?

**Comment by bluebill1049** [comment]
@Flavien-Pensato if your name type is `string | number`, then it should be good right?

**Comment by Flavien-Pensato** [comment]
Ah okay then to be subject of rules validation I have to declare GlobalState right ? (I don't use typescript)

**Comment by bluebill1049** [comment]
haha yea @Flavien-Pensato 🙏

File: src/types.ts [hunk 5]
```diff
@@ -1,18 +1,17 @@
...
-
-export type Action<T> = (payload: DeepPartial<T>) => void;
+export type AnyCallback = (state: GlobalState, payload: any) => GlobalState
 
-export type Actions<T, K> = Record<keyof K, Action<T>>;
+export type AnyActions<TCallback> = Record<string, TCallback>;
 
-export type ActionArg<T> = (globalStore: T, payload: DeepPartial<T>) => T;
+export type ActionsOutput<TCallback extends AnyCallback, TActions extends AnyActions<TCallback>> = {
+  [K in keyof TActions]: (payload: Parameters<TActions[K]>[1]) => void;
```
Comment (bluebill1049) [inline src/types.ts:8]: getting the second argument is awesome!!

**Review by bluebill1049 — APPROVED** [review]
Thank you so much for this amazing PR! YAY type-safe v4!

File: src/index.ts (hunk 2 above)
Comment (kotarella1110) [reply src/index.ts:8]: 👄

**Comment by kotarella1110** [comment]
@bluebill1049 @devrnt @Flavien-Pensato
example here: https://codesandbox.io/s/typescript-s7u3w?file=/src/App.tsx

**Pull request merged**